pkg/pillar/cmd/zedmanager/memorysizemgmt.go

```diff
@@ -29,6 +29,7 @@ func getRemainingMemory(ctxPtr *zedmanagerContext) (uint64, uint64, uint64, erro
 	for _, st := range itemsAppInstanceStatus {
 		status := st.(types.AppInstanceStatus)
 		mem := uint64(status.FixedResources.Memory) << 10
+		mem += status.MemOverhead
```

Comment:
I would go further with the proposed change. The 'haltingMemorySize' is incorrect and dangerously racy: an app is marked as "Active = false", but qemu process is still alive (takes time to bring it down), but we already start allocating memory for a new app, considering that memory should be enough. That's bad. I propose avoid the 'halting' calculations at all, and always account "vRAM + overhead" memory for all existent apps.

Reply:
Also the `sysTotalMemory()` (used for `deviceMemorySize`) seems not very reliable: asynchronous pub/sub with 25 seconds period (see `metricsTimerStart()`). In 25 seconds a couple of fat Vms can be started, but calculated `deviceMemorySize` won't account this, which is again bad. Sys memory calculations should very simple and reliable: request true sys  memory counters just from the code accessing the /proc synchronously, that will guarantee that we see current state of the system

Reply:
> I propose avoid the 'halting' calculations at all, and always account "vRAM + overhead" memory for all existent apps.

For me, it makes sense. I'll recheck how this halting memory is used and why it's separated from the "occupied" memory. Maybe @eriknordmark can give some clarifications here.

> Sys memory calculations should very simple and reliable: request true sys memory counters just from the code accessing the /proc synchronously, that will guarantee that we see current state of the system

I have not checked the implementation of the sysTotalMemory() yet. Will do. But your suggestion looks meaningful to me.

Reply:
@OhmSpectator I didn't even know there was a separate halting size. What does git blame say abouts its history?

We do want to trigger a retry  (of apps which couldn't started due to lack of memory) when an app has halted, but that doesn't require an exact number, and as Roman said, the retry can't be done until it is halted.

Reply:
@eriknordmark actually, it was added by you =D
https://github.com/lf-edge/eve/commit/51ffeb1316727a3ad1b444f39adff29a068f6ebb
I'll try to understand the concept now.

Reply:
Here is my current understanding of the different "memories" used when when we decide wether ti run a new domain:

### usedMemorySize

**Description**: This represents the total memory used by applications that are either activated or in the process of being activated.

**Details**: It's a cumulative measure of the memory requirements of all running or activating applications. This includes both the RAM allocated to each application (`FixedResources.Memory`) and any additional overhead (`MemOverhead`).

### haltingMemorySize

**Description**: A subset of `usedMemorySize`, representing the memory used by applications that are currently in the process of halting.

**Details**: This memory is allocated to applications that have been initiated for shutdown but haven't completed the process yet. Although part of `usedMemorySize`, it's distinguished because this memory is expected to be freed up shortly as the applications complete their halting process.

### latentMemorySize

**Description**: Memory used by applications that are halted, meaning they are neither running nor in the process of activating.

**Details**: This category covers applications that are in a passive state, not actively using resources but still reserving memory. This memory isn't immediately available but may become so if these applications are completely shut down or remain inactive.

### memoryReservedForEve

**Description**: Memory reserved specifically for the EVE system itself.

**Details**: This includes the memory limit set globally for EVE (`EveMemoryLimitInBytes`). If the system uses ZFS for persistence (`PersistZFS`), it also includes the maximum size of the ZFS ARC (`zfsArcMaxLimit`), which is an additional memory reservation for file system caching.

### deviceMemorySize

**Description**: The total memory available on the device.

**Details**: This represents the physical memory capacity of the device. It corresponds to the `TotalMemoryMB` metric sent from the DomainManager and is equivalent to the `MemTotal` record found in `/proc/meminfo` on the device.

### remainingMemory

**Description**: The amount of memory available for new applications or processes.

**Details**: Calculated as the total device memory (`deviceMemorySize`) minus the total used memory (sum of `usedMemorySize` and `memoryReservedForEve`). It represents the current available memory capacity that can be allocated to new or additional applications.

Reply:
And here is how we decide whether to activate a domain or not.

### 1. Calculate Memory Requirements
- **Determine Total Need (`need`)**: Calculate the total memory requirement for the new domain, incorporating fixed memory allocation (`FixedResources.Memory`) and additional overhead (`MemOverhead`).

### 2. Determine Available Memory
- **Calculate `remainingMemory`**: This is the total device memory (`deviceMemorySize`) minus the used memory (`usedMemorySize`) and memory reserved for EVE (`memoryReservedForEve`). Note that halting memory is not considered free here. But latent - is. 

### 3. Decision Making
- **Sufficient Memory Without Halting Memory**: 
  - If `remainingMemory` ≥ `need`, there's enough memory to start the new domain.
  - Warning for potential overcommitment is logged if `remainingMemory` < `need` + `latentMemorySize`. **I find this current behaviour dangerous as it risks system stability by potentially overloading memory resources.** 

- **Potential Sufficiency with Halting Memory**: 
  - If `remainingMemory` < `need`, but `remainingMemory` + `haltingMemorySize` ≥ `need`, delay the start of the new domain.
  - Log a notice, set error severity to notice, and specify `retryCondition` for delayed start.

- **Insufficient Memory Even with Halting Memory (With Running Apps)**: 
  - This scenario occurs when `remainingMemory` + `haltingMemorySize` < `need`, but other applications are running.
  - The presence of running applications suggests potential future availability of memory, leading to a less critical view of the current memory shortage.
  - Set error severity to warning, indicating a non-critical issue that might resolve with the completion of running applications.
  - Delay the start of the new domain and specify `retryCondition` indicating a retry when memory becomes available.

- **Insufficient Memory Even with Halting Memory (No Running Apps)**: 
  - If `remainingMemory` + `haltingMemorySize` < `need` and no applications are running, set the error severity to error.
  - Delay the start of the new domain and specify `retryCondition` for when more memory is available.

Reply:
@eriknordmark, @rouming, my current proposal is:

Step one. We leave the current PR as it is, which adds memOverhead to the code and can potentially prevent the issue of starting multiple large virtual machines that we are currently experiencing.

Step two. We address the underlying issue of latent memory usage in a separate PR. This may require further discussion as it is a more complex problem and won't immediately solve any of the existing issues we are facing.

Reply:
> is equivalent to the `MemTotal` record found in `/proc/meminfo` on the device.

If you plan to use your explanation as documentation, then the above needs more text. If Kvm is is correct, but if Xen then /proc/meminfo is only what is available in dom0 and EVE uses xen tools specific things to find the total memory of the device.

Reply:
> @eriknordmark, actually, it was added by you here [51ffeb1](https://github.com/lf-edge/eve/commit/51ffeb1316727a3ad1b444f39adff29a068f6ebb) =D I'll try to understand the concept now.

Thanks for tracking that down.
Looks like haltingMemorySize is only used to product a slightly different error, hence it does not need to be that accurate. We will trigger a retry when memory when the app instance has halted no matter what.

Reply:
I put this explanation just to clarify the situation and to decide on the following steps with the PR: whether any extra actions were critical for addressing the original issue we were trying to resolve. Looks like they are not, but we still can make improvements. I'll create a ticket and put the comment there. Later I can rework it and make a part of the documentation.

Reply:
Great.